### ISSUE TITLE
Pyright does not respect `__class_getitem__` annotations.

### ISSUE THREAD

**Issue by imgurbot12** [microsoft/pyright#11561]
**Describe the bug**

Pyright does not seem to respect type-annotations defined by custom `__class_getitem__` implementations.

It assumes the getitem will return the associated class type regardless of the annotations present.

**Code or Screenshots**

```python
from typing import Annotated, TypeVar, cast, reveal_type

T = TypeVar('T')

class Foo:
    @classmethod
    def __class_getitem__(cls, type: T) -> T:
        return cast(T, Annotated[type, cls()])

foo = Foo[str]
reveal_type(foo) # Type of "foo" is "type[Foo]"
```

In the example above the type should report as `str`.

**VS Code extension or command-line**

This issue exists both in LSP and with the latest version of the pyright command-line `v1.1.411`

**Comment by imgurbot12** [comment]
Interestingly enough, it works with a metaclass `__getitem__` but not with a standard `__class_getitem__` but only sometimes since it raises an error if the type annotation is defined inside a function:

```python
from typing import Annotated, TypeVar, cast, get_args, reveal_type

T = TypeVar('T')

class Meta(type):
    def __getitem__(self, type: T) -> T:
        return cast(T, Annotated[type, self()])

class Foo(metaclass=Meta):
    pass

foo = Foo[str]
reveal_type(foo) # Type of "foo" is "type[str]"

def bar(a: Foo[int]): # error: Expected no type arguments for class "Foo"
    pass
```

**Comment by imgurbot12** [comment]
Same if you declare a typealias, which I assume is the same error when defining it inside a function.

```python
Test: TypeAlias = Foo[str] # error: Expected no type arguments for class "Foo"
```

**Comment by jorenham** [comment]
Related discussion: https://discuss.python.org/t/should-non-generic-classes-with-class-getitem-be-subscriptable-in-a-type-expression/108666
